File: src/Ara.Application/Common/Validation/PasswordPolicy.cs
using System.Text.RegularExpressions;

namespace Ara.Application.Common.Validation;

/// <summary>
/// Validates a raw (pre-hash) password. Enforce this at registration/password-change time,
/// before the password is hashed — a hash itself can't carry these characteristics.
/// </summary>
public static partial class PasswordPolicy
{
    public const string Requirements =
        "Password must be at least 12 characters and include a number, an uppercase letter, and a special character.";

    public static bool IsValid(string password) => PasswordPattern().IsMatch(password);

    [GeneratedRegex(@"^(?=.*[0-9])(?=.*[A-Z])(?=.*[^A-Za-z0-9]).{12,}$")]
    private static partial Regex PasswordPattern();
}

File: Ara.Client/src/app/core/validation/password-policy.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

// Mirrors Ara.Application.Common.Validation.PasswordPolicy on the backend —
// keep both in sync if this ever changes.
export const PASSWORD_REQUIREMENTS =
  'At least 12 characters, with a number, an uppercase letter, and a special character.';

const PASSWORD_PATTERN = /^(?=.*[0-9])(?=.*[A-Z])(?=.*[^A-Za-z0-9]).{12,}$/;

export function isPasswordValid(password: string): boolean {
  return PASSWORD_PATTERN.test(password);
}

export const passwordComplexityValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
  const value = control.value as string;
  return !value || isPasswordValid(value) ? null : { passwordComplexity: true };
};
